Use v1.Container in Debug Containers API

verb opened the pull request:
This proposes an API for Debug Containers that's idiomatic for Kubernetes and uses the existing `Container` definition. This makes it easier for admission controllers to gate these updates.

The previous Debug Containers proposal avoided changing the core API as much as possible.

verb commented:
/sig node auth
/assign @lavalamp @liggitt @thockin

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
````diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+}
+```
+
+Note that Debug Containers are not regular containers and should not be used to
````
s/debug/ephemeral/g, based on where the review of the last PR ended up?

lavalamp replied:
+1

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+will not be allowed for Debug Containers. A request will be rejected if any
+field is set other than the following whitelisted fields: `Name`, `Image`,
+`Command`, `Args`, `WorkingDir`, `Env`, `EnvFrom`, `ImagePullPolicy`,
+`SecurityContext`. `TTY` and `Stdin` are always enabled for Debug Containers and
```
Running an ephemeral container with non-interactive diagnostics and viewing the log would not require tty/stdin. Are there any containers where enabling tty/stdin would cause problems?

lavalamp replied:
Yeah I don't see why you'd mandate those features on?

verb replied:
iirc this came from a conversation from @thockin and @vishh about how having configurable stdin/tty has caused confusion for users and been more trouble than it was worth. I don't feel strongly.

Some containers don't require tty/stdin, but is there a downside to them being always on? This is a new feature, so there aren't backwards compatibility concerns here.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+We can either using an existing subresource like `/exec` and have the call
+stream or not based on `PodExecOptions`, or we can create a new subresource that
+never streams. For the sake of consistency, we will create a new subresource
+named `/debug`. This has the added benefit of being able to entirely hide
```
A separate subresource makes a lot of sense to me. Adding a newly defined container (new image, new security settings, etc) is very different from exec, defining the API we want for it seems like a good idea.

lavalamp replied:
`/execEphemeralContainer` -- anything other than `/debug`.

A new endpoint seems good for the super-security conscious at the expense of making everything more complicated for everyone else. I don't have a strong opinion.

liggitt replied:
> /execEphemeralContainer -- anything other than /debug.

probably want lowercase resources, but yes

> A new endpoint seems good for the super-security conscious at the expense of making everything more complicated for everyone else. I don't have a strong opinion.

it's good for isolating the function for policy purposes, but also lets us support envvars, capabilities, and other useful knobs for shaping the environment of the ephemeral container in a way consistent with existing pods

lavalamp replied:
> probably want lowercase resources

er, right

> it's good for isolating the function for policy purposes, but also lets us support envvars, capabilities, and other useful knobs for shaping the environment of the ephemeral container in a way consistent with existing pods

I stand by my previous response :)

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+1.  A `POST` of a `PodDebugContainer` to
+    `/api/v1/namespaces/$NS/pods/$POD_NAME/debug` to create a Debug Container
+    running in pod `$POD_NAME`.
+1.  A `DELETE` of `/api/v1/namespaces/$NS/pods/$POD_NAME/debug/$NAME` will stop
```
Named subresources are a thing I have wanted, but not something we currently have, which could make this tricky. I'd see this as nice-to-have, but not essential (as noted in the previous PR, things like exec sessions are not currently able to be individually terminated via the API)

lavalamp replied:
I suggest allowing only one ephemeral container at a time for the moment, and then you don't need the additional level? We don't support this at the moment.

liggitt replied:
alternately, `DELETE` of `/api/v1/namespaces/$NS/pods/$POD_NAME/ephemeralcontainers` stops all ephemeral containers

thockin replied:
Agree we can live without it for now

verb replied:
SGTM, I will omit it for now.

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+`SecurityContext`. `TTY` and `Stdin` are always enabled for Debug Containers and
+will be ignored.
+
+The new `/debug` subresource allows the following:
```
s/`/debug`/something else/g

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+the existing attach endpoint, `/api/v1/namespaces/$NS/pods/$POD_NAME/attach`.
+Note that any output of the new container between its creation and subsequent
+attach will not be replayed and can only be viewed using `kubectl log`.
+
```
Will the ephemeral container appear in the response when I GET the pod?

liggitt replied:
only the status of it, not the spec

lavalamp replied:
That's fine.

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -441,10 +448,11 @@ The kubelet exposes the new functionality in its existing `/exec/` endpoint.
 `RunDebugContainer()`, and performs the attach.
 
-##### Step 4: Reporting DebugStatus
+##### Step 4: Reporting EphemeralContainerStatus
```
Ah, looks like the answer is yes.

lavalamp commented:
This LGTM to me if you can fix the naming (consistency + debug -> ephemeral).

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -235,37 +249,28 @@ The status of a Debug Container is reported in a new field in `v1.PodStatus`:
-there is no spec for Debug Containers or exec. These must either be made
+Note that `Command` and `Args` would have to be tracked in the status object
+because there is no spec for Debug Containers or exec. These must either be made
 available by the runtime or tracked by the kubelet. For Debug Containers this
```
doesn't the whole ephemeral container have to be tracked by the kubelet?

liggitt replied:
why just the command and args? admission uses the definitions of containers to determine whether a user can do things like exec/attach to those containers ("cannot exec into a pod with containers running as root", etc). Container status alone (`EphemeralContainerStatuses []v1.ContainerStatus`) wouldn't include enough information about the ephemeral containers to make that determination.

what if the submitted ephemeral container and its status were both made available in pod status?

```
type PodStatus struct {
  ...
  EphemeralContainers []v1.EphemeralContainer
}

type EphemeralContainer struct {
  Spec v1.Container
  Status v1.ContainerStatus
}
```

verb replied:
The kubelet currently has no method of keeping state of the v1.Container across restarts. After a restart, it can reconstruct a minimum v1.Container based on runtime labels, though. This could (but currently doesn't) include Command and Args, which can be queried from the run time.

Including the spec in the PodStatus would be pretty great, though, so it might be worth solving. @lavalamp what do you think about this?

lavalamp replied:
Remind me why this got reopened for discussion? I feel like if we keep going down this route we will end up at my original position of adding a `type` of `ephemeral` to containers, and allowing users to add or remove such containers from the spec.

verb replied:
I think sig-auth just wants to make this works well with admission controllers and has a strong preference for using v1.Container, [for many good reasons](https://github.com/kubernetes/community/pull/649/files#r142722579).

@liggitt This is an important feature to sig-node, but we've been stuck for about a year. If a v1.Container is available to admission controllers for the creation request, can you live with only ContainerStatus for attach, at least for alpha? We don't need to allow exec.

liggitt replied:
> The kubelet currently has no method of keeping state of the v1.Container across restarts.

I expected the kubelet to:
1. update pod status with the ephemeral container definition/status as it created the container
2. update the ephemeral container status by name from observed container state during normal sync
3. remove the ephemeral container from the pod status when the ephemeral container was removed

None of that requires the kubelet remembering the ephemeral container definition beyond the initial create, right?

> If a v1.Container is available to admission controllers for the creation request, can you live with only ContainerStatus for attach, at least for alpha? We don't need to allow exec.

Two thoughts:
1. attach (which is the intended use case here) requires almost identical protections to exec ("cannot attach to a pod with containers running as root", etc). If the create of the ephemeral container and the attach to it are separate API calls, but the container definition is unavailable at attach time, I'm unsure how that would be implementable.
2. the structure of an ephemeral container field in the pod status could not be changed [in an incompatible way](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions) (*if the schema of the alpha field must change in an incompatible way, a new field name must be used*), so if we started with `EphemeralContainerStatuses []v1.ContainerStatus`, that couldn't grow to include the container definition as alpha progressed.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+build services. They have no guarantees and most of the fields of `v1.Container`
+will not be allowed for Debug Containers. A request will be rejected if any
+field is set other than the following whitelisted fields: `Name`, `Image`,
+`Command`, `Args`, `WorkingDir`, `Env`, `EnvFrom`, `ImagePullPolicy`,
```
are HostNetwork/HostPID/HostIPC inherited from the pod spec?

liggitt replied:
should also document this in a way that doesn't preclude supporting new container fields that make sense in the future

lavalamp replied:
Yeah, I think it's a possibility that we'll want to support more of these fields as time goes on.

verb replied:
Actually, the namespaces will be inherited from the target container. Originally it was going to be inherited from the pod spec. This is something that changed a few weeks ago because we're now going to support isolated PID namespaces indefinitely, and it's not reflected in this API alternative.

With exec++ the target container was provided in PodExecOptions, but switching the API means we either have to provide the target in the URI, the posted EphemeralContainer or... query params? I haven't looked into this at all yet.

thockin commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
````diff
@@ -106,13 +107,63 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+```
+// EphemeralContainer describes a container to attach to a running pod for troubleshooting.
+type EphemeralContainer struct {
+        metav1.TypeMeta
````
I assume the name has to be unique within the pod?

lavalamp replied:
Yes, but the container name would be embedded in the ContainerSpec.

This type probably doesn't need an ObjectMeta--I would try it that way and see if any of the encoding stack gives you a problem.

verb replied:
Yes, it must be unique within the pod because it will be used for attach, logs, etc. Thanks for the tip on ObjectMeta, I will try it.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -235,37 +249,28 @@ The status of a Debug Container is reported in a new field in `v1.PodStatus`:
 could be stored as runtime labels, but the kubelet currently has no method of
 storing state across restarts for exec. Solving this problem for exec is out of
 scope for Debug Containers, but we will look for a solution as we implement this
 feature.
```
@smarterclayton was there something were you mentioning that allowed connecting data to running containers?

smarterclayton commented:
I'd like to get this into the alpha slot for 1.10.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +101,76 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+        // If specified, the ephemeral container will be started in the namespaces
+        // of this TargetContainerName. If omitted the container will inherit the
+        // behavior of the pod for namespaces.
+        TargetContainerName string `json:"spec" ...`
```
copy/pasta, `json:"targetContainerName"`

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -437,14 +453,15 @@ Containers unless the sandbox is restarted.
 ##### Step 3: kubelet API changes
 
-The kubelet exposes the new functionality in its existing `/exec/` endpoint.
+The kubelet exposes the new functionality in a new `/podDebug/` endpoint.
```
needs updating to indicate it is given the v1.Container and optional target container name

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -387,14 +407,10 @@ basic functionality:
+default. The following are explicitly out of scope for the 1.10 alpha release:
 
 *   Exited Debug Containers will be garbage collected as regular containers and
     may disappear from the list of Debug Container Statuses.
```
> Exited Debug Containers will be garbage collected as regular containers and may disappear from the list of Debug Container Statuses.

harmonize this with the "The list of `EphemeralContainerStatuses` is never truncated." statement below

liggitt commented:
> I'd like to get this into the alpha slot for 1.10.

Agree. API-wise, this seems to have the info I would expect now (v1.Container in debug request, spec and status persisted to allow gating attach requests).

More detail is needed for information flow and sequence (which component persists which bits, synchronously or asynchronously,  etc) and some edge cases (what happens on kubelet restart? kubelet+docker restart?), but that can be a further iteration.

verb commented:
@liggitt awesome, thanks for the quick feedback. Since this looks to be heading in the right direction for the API, I'll double check it's still acceptable to sig-node and then do a more complete rewrite addressing your comments.

verb commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -192,104 +225,86 @@ type EphemeralContainer struct {
+it supports up to the 64K constraint of the docker client's `bufio.Scanner`
+size. Because the container spec may be examined in security sensitive contexts
+like admission control, we will conservatively limit the size of the spec to 32K
+and add a 32K minimum label length test to runtime qualification.
```
@mrunalp This is the section where I'd appreciate your review. I'm planning on storing the v1.Container in the runtime. This works for Docker. Do you know of any limitations for cri-o?

verb commented:
Swapping back in after focusing on kubernetes/features#495. I'll be creating PRs for the features described here over the next week.

I've gone through the doc to make it consistent with our recent changes, and I've added more details about how the kubelet will start the container and avoid races.

@liggitt Do these changes sound good enough (at least for alpha) that I can ask the sig-node-leads to review and approve this PR?

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +101,88 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+`EnvFrom`, `ImagePullPolicy`, `SecurityContext`. `TTY` and `Stdin` are always
+enabled for Debug Containers and will be ignored.
+
+The new `/add` subresource allows the following:
```
is the subresource `/add` or `/ephemeralContainers`? being more explicit seems better (though we don't mix case in resource or subresource names)

verb replied:
It's `/ephemeralcontainers` (I discovered the problem with case on subresources today, too). Fixed.

The kubelet API seems to still mix case, so I left that one alone. sig-node probably has a preference but I'm not sure what it is.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -192,96 +225,100 @@ type EphemeralContainer struct {
-implicit attach maintains consistent semantics across `/exec` rather than
-varying behavior based on parameters.
+    validation, admission control and proxies the connection to the kubelet's
+    `/ephemeralContainers/$NS/$POD_NAME` endpoint.
```
it would be helpful to clarify at what point this call completes (e.g. does it block on the syncPod call or return immediately)

liggitt replied:
nm, I see the details below

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -192,96 +225,100 @@ type EphemeralContainer struct {
 
-1.  `kubectl` invokes the exec API as described in the preceding section.
+1.  `kubectl` invokes the new API as described in the preceding section.
 1.  The API server checks for name collisions with existing containers, performs
```
I'm assuming the kubelet also checks for name collisions with existing containers or existing debug containers, and that the API server validation prevents name collisions when the kubelet updates pod status with ephemeral container status

verb replied:
That's correct. Checking for collisions in the api server provides a quick, user-friendly error while the kubelet resolves any races in its dedicated pod worker.

I've updated this section to clarify.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -192,96 +225,100 @@ type EphemeralContainer struct {
+    `UpdatePodOptions.OnCompleteFunc`.
+    1.  The pod sync worker runs the existing `syncPod()` with a new
+        `SyncPodType` of `SyncPodDebug`.
+    1.  The request handler blocks (with timeout) on receiving an error from
```
what steps occur within this timeout? pulling the image? how is the timeout selected here?

verb replied:
Yes, the container is started, including pulling the image if necessary. I was just going to make the timeout flag configurable. I'd prefer to improve the reliability of the `OnCompleteFunc` for `SyncPodDebug` to making the timeout more configurable. I've described this a little bit in the alpha notes. It should be easy to get to the point where the timeout is rarely used.

verb replied:
Oh, also I updated the doc to reflect this.

liggitt replied:
I'm trying to determine whether an API request to add an ephemeral container should be considered a long running request on the API side. Pulling the image within the lifetime of an API request is a new direction here.

Also, if pulling happens prior to the kubelet persisting the ephemeral container spec in the pod status, does that mean the kubelet gets a single shot at pulling and never reattempts or reports status on failure?

verb replied:
Oh, interesting. Yes, by design we want a failure in starting the ephemeral container to be a permanent failure. There will never be any reattempts, but it should report the failure back to the user, and in the pod's event log.

For example, image pull failures return a ErrImagePull to the user. I'm not sure of the current kubelet behavior for a really slow image pull, but this would be the same for the Ephemeral Container since it uses the same code.

So if len(image_pull_secs) > len(api_request_timeout_secs) then the user will get error("start ephemeral container timed out") even though the ephemeral container will eventually be created (or not). If the Ephemeral Container is eventually created, it will be available for manual attach.

liggitt replied:
> by design we want a failure in starting the ephemeral container to be a permanent failure

that seems fine

> There will never be any reattempts, but it should report the failure back to the user

I'm having difficulty understanding how this works when combined with the image pull happening within the timeout. Would a pull failure still result in a pod status update? Would it make more sense for the subresource POST to just update pod status with the requested ephemeral container, the kubelet notice that via normal pod watch events, and update pod status with the results of attempting to start the ephemeral container?

That would allow the client side implementation to look like:
1. submit ephemeral subresource request, get back immediate success or error
2. on success, monitor pod status until ephemeral container is running
3. attach

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -336,13 +373,13 @@ It's worth noting some things that do not change:
-parameter as is enforced for regular containers. During the alpha phase we will
-additionally support a container image whitelist as a kubelet flag to allow
-cluster administrators to easily constraint debug container images.
+Admission plugins must be updated to guard `/debug`. In particular, they should
```
fix the subresource name

verb replied:
fixed.

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -336,13 +373,13 @@ It's worth noting some things that do not change:
+enforce the same container image policy on the `Image` parameter as is enforced
+for regular containers. During the alpha phase we will additionally support a
+container image whitelist as a kubelet flag to allow cluster administrators to
+easily constraint debug container images.
```
typo: constrain

verb replied:
fixed

liggitt commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -356,38 +393,47 @@ cluster administrators to easily constraint debug container images.
+    proxied to the kubelet.
+1.  The API server must check for Ephemeral Containers when validating `attach`.
+
+New fields will be prepended with `alpha` for the duration of this feature's
```
remove this... field name prefixing is not recommended in https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions

verb replied:
Fixed.

liggitt commented:
thanks for the update. a couple nits, lgtm otherwise.

yujuhong commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -356,38 +403,44 @@ cluster administrators to easily constraint debug container images.
-    be run with `CAP_SYS_PTRACE` and `CAP_SYS_ADMIN`.
-*   Image pull policy for the Debug Container is not configurable. It will
-    always be run with `PullAlways`.
+*   There's no guarantee that exited Debug Containers won't be garbage collected
```
What's the expected normal garbage collection behavior? Do debug containers get removed as regular containers?

verb replied:
Last I checked the garbage collector kept the most recent exited container, which would be sufficient for most purposes. I plan on providing knob that cluster admins can use to exempt debug containers from garbage collection for the life of the pod as this was a point of contention for sig node, but before doing so I want to reopen the discussion to determine the exact requirement.

yujuhong commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -356,38 +403,44 @@ cluster administrators to easily constraint debug container images.
-*   Image pull policy for the Debug Container is not configurable. It will
-    always be run with `PullAlways`.
+*   There's no guarantee that exited Debug Containers won't be garbage collected
+    as regular containers, so they may disappear from the list of
```
Is kubelet supposed to continue reporting status for exited debug containers?

verb replied:
Yes, it should report the most recent container by name, regardless of status.

verb commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -192,96 +225,110 @@ type EphemeralContainer struct {
+    sync worker with `UpdatePodOptions` that include the `EphemeralContainer` in
+    a new field and a callback in the existing
+    `UpdatePodOptions.OnCompleteFunc`.
+    1.  The pod sync worker runs the existing `syncPod()` with a new
```
The debug containers won't be sync'd, except to update ContainerStatus. After being started it will be ignored until the pod sandbox dies (or is garbage collected). Clarified.

verb commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -192,96 +225,110 @@ type EphemeralContainer struct {
+        `syncPod()` via the callback. During this time, `syncPod()` starts the
+        ephemeral container, including fetching an image if necessary, and
+        publishes a new `PodStatus`.
+    1.  Timeout is configured by the cluster administrator and defaults to 15
```
No? I'm not sure what to expect for image pulls. Maybe we could set a long (~2 min) timeout and use a ticker to periodically check the PodStatus cache to see if the container was created but `OnCompleteFunc` dropped...

In any case this feels like something on which we can iterate, but you've brought up a really good point that we should initially err on the side of too long. Let's start with 2 minutes.

verb commented:
I've updated this with a change suggested by @liggitt where the client updates the `EphemeralContainer` via the API and the ephemeral container is created during the regular kubelet `syncPod`

yujuhong commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -356,38 +358,44 @@ cluster administrators to easily constraint debug container images.
+when Debug Containers are garbage collected. For the alpha release we will rely
+on the apiserver to store the `EphemeralContainer` for garbage collected
+containers. The kubelet will preserve any `EphemeralContainer` it doesn't
+recognize when updating status. In the event that a `PodStatus` is lost and we
```
In what case would `PodStatus` be lost...?

yujuhong commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +101,88 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+debug container using the existing attach endpoint,
+`/api/v1/namespaces/$NS/pods/$POD_NAME/attach`. Note that any output of the new
+container between its creation and subsequent attach will not be replayed and
+can only be viewed using `kubectl log`.
```
Does this mean `kubectl log` will be supported for debug containers and is it treated as a regular container? What other commands can be performed against a debug container, e.g., `exec`? I think it's worth calling out what operations will be supported for debug containers.

verb commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +100,91 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+        // Populated by the system.
+        // Read-only.
+        // +optional
+        Status *ContainerStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```
I'm second guessing my decision to embed the `ContainerStatus` into `EphemeralContainer` as I'm polishing up the kubelet implementation where things just expect to see a `[]ContainerStatus`.  I wonder if it would be better to do something like:

```
type EphemeralContainer struct {
    ...
    Spec *Container
    TargetContainerName string
}

type PodStatus struct {
   ...
   EphemeralContainers []EphemeralContainer
   EphemeralContainerStatuses []ContainerStatus
}
```

I guess that's just pushing the complexity of merging by name onto the client, though, where presumably the spec & status will frequently be operated upon in tandem. If targeting were unnecessary (or could be specified in `Container`) then we wouldn't need the `type EphemeralContainer` at all.

Do you have any suggestions @yujuhong or @liggitt ?

yujuhong replied:
I think the originally proposed `EphemeralContainer` struct that groups everything together might be slightly better to parse.

yujuhong commented:
/lgtm

/assign @dchen1107 @derekwaynecarr 
for approval

@liggitt @lavalamp @thockin do you still have any concern?

dchen1107 commented:
/lgtm

liggitt commented:
> @liggitt @lavalamp @thockin do you still have any concern?

this describes an approach that ensures:
* the information needed for authz/admission/audit is present in the API requests from the user
* the persisted API object captures what ephemeral container(s) were run
* the ephemeral containers are described using v1.Container structs
so my concerns were addressed

thockin commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
````diff
@@ -106,13 +100,91 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+```
+// EphemeralContainer describes a container to attach to a running pod for troubleshooting.
+type EphemeralContainer struct {
+        metav1.TypeMeta `json:",inline"`
````
This makes me think it is a top-level resource - no other sub-structs have their own `kind` fields - why this one?  Can you make sure to explain it in comments, if it is intended?

lavalamp replied:
It is a subresource, like scale.

I don't think this can support anything but POST, is that the intention?

lavalamp replied:
Should we include metadata, and force the user to copy the UID of the pod they are targeting, to prevent accidentally getting a different pod than they intended if there is a delete/recreate?

smarterclayton replied:
We definitely need preconditions to be possible here.  Requiring resourceVersion and uid has been sufficient for all other apis so far.

thockin commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +100,91 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+        // +optional
+        Status *ContainerStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+
+        // If set, the name of the container from PodSpec that this ephemeral container targets.
```
I don't understand what "targets" means.  How can a container target a container?

verb replied:
(and I'll update the comment to make this clear)

thockin commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +100,91 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+the desired `v1.Container` as `Spec` in `Pod.Status` and updates the `Pod` in
+the API. Users cannot normally modify the pod status, so we'll create a new
+subresource `/ephemeralcontainers` that allows an update of solely
+`EphemeralContainers` and enforces append-only semantics.
```
if it is append-only does that mean no cleanup is possible?  Can we say that?

lavalamp replied:
Users should not modify status, period. Adding a subresource doesn't remove that requirement. The subresource is so it can be permissioned separately. The user intent still should go in spec somewhere.

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +100,91 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+
+An Ephemeral Container is not part of the pod specification as it's not part of
+the declared state of the pod, but we describe it using the same primitives as
+in `PodSpec`. An `EphemeralContainer` contains a Spec, a Status and a Target:
```
what verbs does the subresource support?

lavalamp commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +100,91 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+        ...
+        // List of user-initiated ephemeral containers that have been run in this pod.
+        // +optional
+        EphemeralContainers []EphemeralContainer `json:"commands,omitempty" protobuf:"bytes,11,rep,name=ephemeralContainers"`
```
I still strongly disagree with listing a container spec anywhere inside pod status.

lavalamp replied:
I also maintain that you cannot use the same type as the subresource for this in any case-- no typemeta/metadata makes sense here.

lavalamp commented:
Sorry, I think I already said these things on the PR. This came up in my email and I thought it was a recent update but now I think it hasn't been updated recently. Sorry for the noise.

vishh commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,13 +100,91 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+container occurring between its creation and attach will not be replayed, but it
+can be viewed using `kubectl log`.
+
+#### Alternative 1: "exec++"
```
@verb can you add a few lines to describe why these alternatives were not chosen at the end?

thockin commented:
/lgtm
/approve

thockin commented:
I think that is a broken resolver and a broken DNS config.  You should
never have a host named localhost, and DNS clients should not try to
resolve it

On Thu, Aug 23, 2018 at 5:23 PM kfox1111 <notifications@github.com> wrote:

> BTW, speaking of from scratch images, I hit a case in tiller the other day
> where using a thin image where the go binary is statically linked, there is
> no /etc/nsswitch.conf, the go resolver will default to 'dns, files' which
> means if your dns search path happens to have a 'localhost.foo.whatever'
> and your search path includes foo.whatever, localhost resolves to
> localhost.foo.whatever's. This is very unexpected/unfortunate. As this is a
> common k8s method of deployment, we might want to address this. Should we
> file an issue in k8s? in go?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/1269#issuecomment-415613012>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVF_t3LOwXtJoBQD4rACpILW12OP3ks5uT0dugaJpZM4QGMMg>
> .
>

kfox1111 commented:
I agree on both accounts. :)

So, its a bug in golangs resolver then and we need to get golang to fix it.

kfox1111 commented:
one other possibility, as the golang devs have known about the issue for a while and have ignored it, would be to have k8s create /etc/nslookup.conf by default in the container with some sane defaults like it does resolv.conf and hosts.

thockin commented:
The problem is that nsswitch (I assume that is what you meant) is an
artifact of glibc's DNS resolver.  Are we willing to manage that file in
perpetuity for all users? Also, this will brek any image that has a custom
nsswitch file built-in.

On Thu, Aug 23, 2018 at 5:38 PM kfox1111 <notifications@github.com> wrote:

> one other possibility, as the golang devs have known about the issue for a
> while and have ignored it, would be to have k8s create /etc/nslookup.conf
> by default in the container with some sane defaults like it does
> resolv.conf and hosts.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/1269#issuecomment-415615242>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVG56Vjf8ClThmlrKZOUCAQlyuF5jks5uT0r9gaJpZM4QGMMg>
> .
>

kfox1111 commented:
yeah, thats what I meant. It would suck to maintain it for ever, but it is unclear if anyone else is willing to fix it. Its been a known issue for a long time apparently.

Is it easy to do it conditionally? check if the file exists, and if not inject it?

Alternately, it should be exceedingly rare I think for that file to be non default.  We could make it default to always override it, and add a flag not too.

thockin commented:
I'm not sure what a good answer here looks like.  It's sort of a broekn
image that depends on it but doesn't provide it.  Can we really be in the
business of fixing broken images dynamically?

Maybe we can instead raise the heat on the Go devs to get a discussion
going?

On Mon, Aug 27, 2018 at 11:02 AM kfox1111 <notifications@github.com> wrote:

> yeah, thats what I meant. It would suck to maintain it for ever, but it is
> unclear if anyone else is willing to fix it. Its been a known issue for a
> long time apparently.
>
> Is it easy to do it conditionally? check if the file exists, and if not
> inject it?
>
> Alternately, it should be exceedingly rare I think for that file to be non
> default. We could make it default to always override it, and add a flag not
> too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/1269#issuecomment-416313888>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVG76CNnPaBIVQzW07qKUL-slcDavks5uVDQZgaJpZM4QGMMg>
> .
>

kfox1111 commented:
the image is not broken in that is provides a libc that never uses it (not glibc). So it doesn't provide a file it isn't going to use. go is assuming it is using glibc and following glibc convention for how to behave when nsswitch is missing. the binary is statically linked though so doesn't actually use libc. The end user is expecting a statically linked binary to work properly on any linux, and is preferring alpine as its smaller and we're encouraging smaller containers. So, its hard to point at any one thing as the thing to fix, but it is unexpectedly effecting users. We can raise it with the go devs again and see what they say.

thockin commented:
Please feel free to loop me into a discussion with go devs.

On Tue, Aug 28, 2018, 8:18 AM kfox1111 <notifications@github.com> wrote:

> the image is not broken in that is provides a libc that never uses it (not
> glibc). So it doesn't provide a file it isn't going to use. go is assuming
> it is using glibc and following glibc convention for how to behave when
> nsswitch is missing. the binary is statically linked though so doesn't
> actually use libc. The end user is expecting a statically linked binary to
> work properly on any linux, and is preferring alpine as its smaller and
> we're encouraging smaller containers. So, its hard to point at any one
> thing as the thing to fix, but it is unexpectedly effecting users. We can
> raise it with the go devs again and see what they say.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/1269#issuecomment-416626572>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVH90U4yWf-ucpEURspPYEbZfFu8nks5uVV8-gaJpZM4QGMMg>
> .
>

krmayankk commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -45,26 +45,25 @@ A solution to troubleshoot arbitrary container images MUST:
 command, `kubectl debug`, which parallels an existing command, `kubectl exec`.
-Whereas `kubectl exec` runs a *process* in a *container*, `kubectl debug` will
-be similar but run a *container* in a *pod*.
+Whereas `kubectl exec` runs a _process_ in a _container_, `kubectl debug` will
```
i think this is underdocumented here https://kubernetes.io/docs/tasks/debug-application-cluster/get-shell-running-container/ that kubectl exec runs a process in an existing container

krmayankk commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -82,22 +81,16 @@ kubectl debug target-pod
-As is already the case with regular containers, Debug Containers can enter
-arbitrary namespaces of another container via `nsenter` when run with
-`CAP_SYS_ADMIN`.
+the pod. When [process namespace sharing](https://features.k8s.io/495) is
```
the top line should make it more clear that  this creates a new container and runs an interactive shell inside that container . `This creates a new container and starts an interactive shell inside that container in a pod called target-pod which can examine and signal other processes in the pod`

krmayankk commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,243 +99,180 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+  // List of user-initiated ephemeral containers to run in this pod.
+  // This field is alpha-level and is only honored by servers that enable the EphemeralContainers feature.
+  // +optional
+  EphemeralContainers []EphemeralContainer `json:"ephemeralContainers,omitempty" protobuf:"bytes,29,opt,name=ephemeralContainers"`
```
how is this different from regular container ?

krmayankk commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,243 +99,180 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+  // If not set then the ephemeral container is run in whatever namespaces are shared
+  // for the pod.
+  // +optional
+  TargetContainerName string `json:"targetContainerName,omitempty" protobuf:"bytes,2,opt,name=targetContainerName"`
```
all containers of the pods are in same network namespace except possibly pid namespace. Was this field introduced for the pid namespace case only ?

krmayankk commented on contributors/design-proposals/node/troubleshoot-running-pods.md:
```diff
@@ -106,243 +99,180 @@ subsequently be used to reattach and is reported by `kubectl describe`.
+Much of the utility of Ephemeral Containers comes from the ability to run a
+container within the PID namespace of another container. `TargetContainerName`
+allows targeting a container that doesn't share its PID namespace with the rest
+of the pod. We must modify the CRI to enable this functionality (see below).
```
So `TargetContainerName` is only useful for pid namespace case only . It would be good to clarify that it has no meaning when all containers are in same network and pid namespace as pod
